src/Ninject.Web.AspNetCore/ServiceCollectionAdapter.cs

```diff
@@ -0,0 +1,103 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+using Ninject.Syntax;
+using Ninject.Web.Common;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Ninject.Web.AspNetCore
+{
+	public class ServiceCollectionAdapter
+	{
+		private readonly IDictionary<Type, BindingIndex> _bindingIndexMap = new Dictionary<Type, BindingIndex>();
+
+		public void Populate(IKernel kernel, IServiceCollection serviceCollection)
+		{
+			if (serviceCollection == null)
+			{
+				throw new ArgumentNullException(nameof(serviceCollection));
+			}
+
+			var adapters = kernel.GetAll<IPopulateAdapter>().ToList();
+
+			foreach (var descriptor in serviceCollection)
+			{
+				if (UsedCustomBinding(kernel, adapters, descriptor))
+				{
+					continue;
+				}
+
+				ConfigureImplementationAndLifecycle(kernel.Bind(descriptor.ServiceType), descriptor);
+			}
+
+			foreach (var adapter in adapters)
+			{
+				adapter.AdaptAfterPopulate(kernel);
+			}
+		}
+
+		private bool UsedCustomBinding(IKernel kernel, List<IPopulateAdapter> adapters, ServiceDescriptor descriptor)
+		{
+			foreach (var adapter in adapters)
+			{
+				if (adapter.AdaptDescriptor(kernel, descriptor))
+				{
+					return true;
+				}
+			}
+
+			return false;
+		}
+
+		private IBindingWithOrOnSyntax<T> ConfigureImplementationAndLifecycle<T>(
+			IBindingToSyntax<T> bindingToSyntax,
+			ServiceDescriptor descriptor) where T : class
+		{
+			if (!_bindingIndexMap.ContainsKey(descriptor.ServiceType))
```

Comment:
This here is the important bit for finding the "latest" binding. Every `BindingIndex` is essentially tied to a `ServiceType` here and for every new binding, a new `BindingIndex.Item` is created that links back to the index "root" so that the metadata of the binding can _know_ what it's own binding index is and if it is the last one.